### ISSUE TITLE
Native assets maybe broken on macOS Sonoma

### ISSUE THREAD

**Issue by craiglabenz** [dart-lang/native#199]
## The problem

Flutter cannot find symbols using `ffi-native:` that it **can** find when the ffi-native feature is disabled and the dylib is opened via

```dart
DynamicLibrary.open('path/to.dylib')
```

### Background information

1. This process worked until I updated my work machine to macOS Sonoma
  1.b. I don't think anything else has changed 
3. I have updated Xcode to 15.0.1 and recompiled
4. Using the latest build from the `master` channel:
```sh
$ flutter doctor -v
[✓] Flutter (Channel master, 3.17.0-9.0.pre.10, on macOS 14.1.1 23B81 darwin-arm64, locale en)
    • Flutter version 3.17.0-9.0.pre.10 on channel master at /Users/craiglabenz/Dev/git/flutter/flutter
    • Upstream repository git@github.com:flutter/flutter.git
    • Framework revision 449908670a (32 minutes ago), 2023-11-14 13:49:26 -0500
    • Engine revision 603bdd48df
    • Dart version 3.3.0 (build 3.3.0-127.0.dev)
    • DevTools version 2.30.0-dev.4

  ...

[✓] Xcode - develop for iOS and macOS (Xcode 15.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15A507
    • CocoaPods version 1.13.0
```
5. The error:
```
Unhandled exception:
Invalid argument(s): Couldn't resolve native function 'text_classifier_create' in 'package:mediapipe_text/src/third_party/mediapipe/generated/mediapipe_text_bindings.dart' : dlsym(RTLD_DEFAULT, text_classifier_create): symbol not found.
```
6. Evidence the symbol exists:
```
$ nm -a libtext_classifier.dylib | grep text_classifier_create
0000000000006488 T _text_classifier_create
```

cc @dcharkes

**Comment by craiglabenz** [comment]
Update: I don't know WHY I bumped into this all of the sudden now, but I realized my `Asset.id` value in `build.dart` had de--synced from the actual location of my bindings.

```dart
Asset(
  id: 'package:mediapipe_text/src/third_party/mediapipe/generated/mediapipe_text_bindings.dart',
  ...
)
```

Changing values to this in my `build.dart` file works. Previously in all of our conversation, it hadn't been obvious to me that this `id` value HAD to be the same import path as the bindings file.

**Comment by mkustermann** [comment]
> Changing values to this in my build.dart file works. Previously in all of our conversation, it hadn't been obvious to me that this id value HAD to be the same import path as the bindings file.

The declarations in Dart code can actually specify an arbitrary asset id - it doesn't have to be the import uri of the library. Maybe our docs / examples should make that explicit. /cc @dcharkes

**Comment by craiglabenz** [comment]
It wasn't working with an `id` value of `package:mediapipe_text/src/mediapipe_text_bindings.dart` 🤔

**Comment by dcharkes** [comment]
> The declarations in Dart code can actually specify an arbitrary asset id - it doesn't have to be the import uri of the library. Maybe our docs / examples should make that explicit. /cc @dcharkes

We do have documentation:

* https://api.dart.dev/stable/3.1.5/dart-ffi/Native-class.html
* https://api.dart.dev/stable/3.1.5/dart-ffi/DefaultAsset-class.html
* I've also added a note in https://github.com/dart-lang/site-www/pull/5347.

**Comment by mkustermann** [comment]
```
Unhandled exception:
Invalid argument(s): Couldn't resolve native function 'text_classifier_create'
in
'package:mediapipe_text/src/third_party/mediapipe/generated/mediapipe_text_bindings.dart' :
dlsym(RTLD_DEFAULT, text_classifier_create): symbol not found.
```

@dcharkes Maybe this error message could instead say:
```
Attempted to lookup `text_classifier_create` in asset <...> failed: No asset with id <...> found (available assets: ...).
Attempted to fallback to process lookup: No symbol in current process.
```

(Pushing for this, because I've also fallen into this problem (of `build.dart` using different asset id than the library) before)

**Comment by dcharkes** [comment]
> ```
> Unhandled exception:
> Invalid argument(s): Couldn't resolve native function 'text_classifier_create'
> in
> 'package:mediapipe_text/src/third_party/mediapipe/generated/mediapipe_text_bindings.dart' :
> dlsym(RTLD_DEFAULT, text_classifier_create): symbol not found.
> ```
> 
> @dcharkes Maybe this error message could instead say:
> 
> ```
> Attempted to lookup `text_classifier_create` in asset <...> failed: No asset with id <...> found (available assets: ...).
> Attempted to fallback to process lookup: No symbol in current process.
> ```

Yeah. I beleive, we used to have a very long error message that would report on all three resolution strategies when resolution failed. But I removed that after getting feedback on the CL. Let's see if I can dig up that code.
